File: src/types/weather.ts
// Core weather data types
export interface Location {
  name: string;
  latitude: number;
  longitude: number;
  country: string;
  region?: string;
}

export interface WeatherData {
  timestamp: string;
  temperature: number; // Celsius
  humidity: number; // Percentage
  pressure: number; // hPa
  windSpeed: number; // m/s
  windDirection: number; // degrees
  cloudCover: number; // Percentage
  precipitation?: number; // mm
  location: Location;
}

export interface UncertaintyBounds {
  lowerBound: number;
  upperBound: number;
  confidence: number; // 0-1
}

export interface ModelPredictions {
  lstm: number;
  arima: number;
  randomForest: number;
  fuzzyTimeSeries: number;
  ensemble: number;
}

export interface PredictionPoint {
  timestamp: string;
  parameter: string;
  value: number;
  uncertainty: UncertaintyBounds;
  modelContributions: ModelPredictions;
}

export interface PredictionResult {
  timestamp: string;
  temperature: number;
  humidity: number;
  pressure: number;
  windSpeed: number;
  uncertainty: UncertaintyBounds;
  modelContributions: ModelPredictions;
}

export interface ModelMetrics {
  mae: number; // Mean Absolute Error
  rmse: number; // Root Mean Square Error
  mape: number; // Mean Absolute Percentage Error
  lastUpdated: string;
}

export interface PredictionResponse {
  location: Location;
  predictionTime: string;
  forecastHorizon: number; // hours
  predictions: PredictionResult[];
  modelPerformance: Record<string, ModelMetrics>;
  currentWeather: WeatherData;
}

export interface CurrentWeatherResponse {
  location: Location;
  current: WeatherData;
  lastUpdated: string;
}

// API Request types
export interface PredictionRequest {
  latitude: number;
  longitude: number;
  hours?: number; // default 24
}

export interface LocationSearchRequest {
  query: string;
  limit?: number;
}

export interface LocationSearchResponse {
  locations: Location[];
}

// Weather parameter types
export type WeatherParameter = 
  | 'temperature' 
  | 'humidity' 
  | 'pressure' 
  | 'windSpeed' 
  | 'cloudCover' 
  | 'precipitation';

// Model types
export type ModelType = 
  | 'lstm' 
  | 'arima' 
  | 'randomForest' 
  | 'fuzzyTimeSeries' 
  | 'lube' 
  | 'ensemble';

export interface ModelStatus {
  name: string;
  type: ModelType;
  isLoaded: boolean;
  lastTrained: string;
  performance: ModelMetrics;
  status: 'healthy' | 'degraded' | 'failed';
}

export interface ModelStatusResponse {
  models: ModelStatus[];
  systemHealth: 'healthy' | 'degraded' | 'failed';
}

// Error types
export interface APIError {
  message: string;
  code: string;
  details?: Record<string, any>;
}

export interface WeatherAPIError extends APIError {
  source: 'openmeteo' | 'weatherapi' | 'openweathermap';
  retryable: boolean;
}

File: backend/app/schemas/weather.py
"""
Pydantic schemas for weather data validation.
"""

from typing import Optional, List
from datetime import datetime
from pydantic import BaseModel, Field, validator


class Location(BaseModel):
    """Location information."""
    name: str
    latitude: float = Field(..., ge=-90, le=90)
    longitude: float = Field(..., ge=-180, le=180)
    country: str
    region: Optional[str] = None


class WeatherData(BaseModel):
    """Current weather data."""
    timestamp: datetime
    temperature: float = Field(..., description="Temperature in Celsius")
    humidity: float = Field(..., ge=0, le=100, description="Humidity percentage")
    pressure: float = Field(..., gt=0, description="Pressure in hPa")
    wind_speed: float = Field(..., ge=0, description="Wind speed in m/s")
    wind_direction: float = Field(..., ge=0, le=360, description="Wind direction in degrees")
    cloud_cover: float = Field(..., ge=0, le=100, description="Cloud cover percentage")
    precipitation: Optional[float] = Field(None, ge=0, description="Precipitation in mm")
    location: Location


class UncertaintyBounds(BaseModel):
    """Uncertainty bounds for predictions."""
    lower_bound: float
    upper_bound: float
    confidence: float = Field(..., ge=0, le=1)


class ModelPredictions(BaseModel):
    """Individual model predictions."""
    lstm: float
    arima: float
    random_forest: float
    fuzzy_time_series: float
    ensemble: float


class PredictionResult(BaseModel):
    """Single prediction result."""
    timestamp: datetime
    temperature: float
    humidity: float
    pressure: float
    wind_speed: float
    uncertainty: UncertaintyBounds
    model_contributions: ModelPredictions


class ModelMetrics(BaseModel):
    """Model performance metrics."""
    mae: float = Field(..., ge=0, description="Mean Absolute Error")
    rmse: float = Field(..., ge=0, description="Root Mean Square Error")
    mape: float = Field(..., ge=0, description="Mean Absolute Percentage Error")
    last_updated: datetime


class PredictionRequest(BaseModel):
    """Request for weather prediction."""
    latitude: float = Field(..., ge=-90, le=90)
    longitude: float = Field(..., ge=-180, le=180)
    hours: int = Field(24, ge=1, le=168, description="Forecast horizon in hours")


class PredictionResponse(BaseModel):
    """Response with weather predictions."""
    location: Location
    prediction_time: datetime
    forecast_horizon: int
    predictions: List[PredictionResult]
    model_performance: dict
    current_weather: WeatherData


class CurrentWeatherResponse(BaseModel):
    """Response with current weather."""
    location: Location
    current: WeatherData
    last_updated: datetime


class LocationSearchRequest(BaseModel):
    """Request for location search."""
    query: str = Field(..., min_length=2, max_length=100)
    limit: int = Field(10, ge=1, le=50)


class LocationSearchResponse(BaseModel):
    """Response with location search results."""
    locations: List[Location]


class APIError(BaseModel):
    """API error response."""
    message: str
    code: str
    details: Optional[dict] = None


class WeatherAPIError(APIError):
    """Weather API specific error."""
    source: str = Field(..., description="API source that failed")
    retryable: bool = Field(True, description="Whether the error is retryable")